knit/src/Knit/Autocompletion.hs

```diff
@@ -0,0 +1,230 @@
+module Knit.Autocompletion where
+
+import Control.Monad.Writer.Strict (runWriter)
+import Data.List (isPrefixOf)
+import Data.Maybe (mapMaybe)
+import Data.Semigroup ((<>))
+import Data.Text (Text, count, drop, dropEnd, pack, replicate)
+import Data.Text.Buildable (build)
+import Data.Text.Lazy (unpack)
+import Data.Text.Lazy.Builder (toLazyText)
+
+import Knit.Argument
+import Knit.FormattedExprExt
+import Knit.Name
+import Knit.Parser
+import Knit.ParseTreeExt
+import Knit.Prelude
+import Knit.Printer
+import Knit.Procedure
+import Knit.Syntax
+import Knit.Tokenizer
+
+suggestions
+  :: forall components proxy.
+     ( KnownSpine components
+     , AllConstrained (ComponentTokenizer components) components
+     , AllConstrained (ComponentTokenToLit components) components
+     , AllConstrained (ComponentCommandProcs components) components
+     , AllConstrained ComponentPrinter components
+     )
+  => proxy components
+  -> Text
+  -> [Text]
+suggestions _ cmd =
+  let
+    openingParens = count "(" cmd
+    closingParens = count ")" cmd
+    addedParens = openingParens - closingParens + 1
+    cmd' = "(" <> cmd <> Data.Text.replicate addedParens ")"
```

Comment:
That's a bad way to do this.

1. Assume the user's input contains a string literal with a parenthesis, e.g. `(foo x:")"`, you will detect as a closing parenthesis, but it isn't.
2. Printing the input string back will now contain parentheses that weren't there in the original user input (I see that you `drop` them a few lines below, but this relies on the fact that the pretty printer does not insert any whitespace between those parentheses - a fragile design).

I asked to modify the grammar to treat EOF as "close all parentheses", didn't this work? Why? Maybe ditch Earley and use parser combinators all the way, then it'll work?

Reply:
1. My bad, indeed.
2. Example from 1 it should fail during parsing not proceed further. In general, I made another printer for `Expr FormattedTreeExt`, the only spaces it prints are spaces which are put in the tree explicitly. But I agree that that design is fragile.
3. I think that I misunderstood you, I didn't try that approach.